File: src/java/3-longest-substr-no-repeating-chars/main.java
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int i = 0,j = 0,result = 0;
        HashSet<Character> visited = new HashSet<>();
        while (i<=j && j<s.length()){
            if (! visited.contains(s.charAt(j))) {
                visited.add(s.charAt(j));
                result = Math.max(result, j-i+1);
                j+=1;
            } else {
                visited.remove(s.charAt(i));
                i+=1;
            }
        }
        return result;
    }
}

File: src/py/3-longest-substr-no-repeating-chars/main.py
class Solution:
    @staticmethod
    def lengthOfLongestSubstring(s: str) -> int:
        result = 0
        i,j = 0,0 
        visited = set()
        while i<=j and j<len(s):
            if s[j] not in visited:
                visited.add(s[j])
                result = max(result, j-i+1)
                j+=1
            else:
                visited.remove(s[i])
                i+=1
        return result

if __name__ == "__main__":
    inputs = [
        ("abcabcbb", 3),
        ("bbbbb", 1),
        ("pwwkew", 3)
    ]
    for s, expected in inputs:
        print(Solution.lengthOfLongestSubstring(s))